src/optimizer/rule/PushFilterDownLeftJoinRule.cpp

```diff
@@ -0,0 +1,130 @@
+/* Copyright (c) 2021 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "optimizer/rule/PushFilterDownLeftJoinRule.h"
+
+#include "optimizer/OptContext.h"
+#include "optimizer/OptGroup.h"
+#include "planner/plan/PlanNode.h"
+#include "planner/plan/Query.h"
+#include "util/ExpressionUtils.h"
+
+using nebula::graph::PlanNode;
+using nebula::graph::QueryContext;
+
+namespace nebula {
+namespace opt {
+
+std::unique_ptr<OptRule> PushFilterDownLeftJoinRule::kInstance =
+    std::unique_ptr<PushFilterDownLeftJoinRule>(new PushFilterDownLeftJoinRule());
+
+PushFilterDownLeftJoinRule::PushFilterDownLeftJoinRule() {
+    RuleSet::QueryRules().addRule(this);
+}
+
+const Pattern& PushFilterDownLeftJoinRule::pattern() const {
+    static Pattern pattern = Pattern::create(graph::PlanNode::Kind::kFilter,
+                                             {Pattern::create(graph::PlanNode::Kind::kLeftJoin)});
+    return pattern;
+}
+
+StatusOr<OptRule::TransformResult> PushFilterDownLeftJoinRule::transform(
+    OptContext* octx,
+    const MatchedResult& matched) const {
+    auto* filterGroupNode = matched.node;
+    auto* oldFilterNode = filterGroupNode->node();
+    auto deps = matched.dependencies;
+    DCHECK_EQ(deps.size(), 1);
+    auto leftJoinGroupNode = deps.front().node;
+    auto* leftJoinNode = leftJoinGroupNode->node();
+    DCHECK_EQ(oldFilterNode->kind(), PlanNode::Kind::kFilter);
+    DCHECK_EQ(leftJoinNode->kind(), PlanNode::Kind::kLeftJoin);
+    auto* oldLeftJoinNode = static_cast<graph::LeftJoin*>(leftJoinNode);
+    const auto* condition = static_cast<graph::Filter*>(oldFilterNode)->condition();
+    DCHECK(condition);
+    const std::pair<std::string, int64_t>& leftVar = oldLeftJoinNode->leftVar();
+    auto symTable = octx->qctx()->symTable();
+    std::vector<std::string> leftVarColNames = symTable->getVar(leftVar.first)->colNames;
+    auto objPool = octx->qctx()->objPool();
+
+    // split the `condition` based on whether the varPropExpr comes from the left child
+    auto picker = [&leftVarColNames](const Expression* e) -> bool {
+        auto varProps = graph::ExpressionUtils::collectAll(e, {Expression::Kind::kVarProperty});
+        std::vector<std::string> propNames;
+        for (auto* expr : varProps) {
+            DCHECK(expr->kind() == Expression::Kind::kVarProperty);
+            propNames.emplace_back(*static_cast<const VariablePropertyExpression*>(expr)->prop());
+        }
+        for (auto prop : propNames) {
+            if (std::find_if(
+                    leftVarColNames.begin(), leftVarColNames.end(), [&prop](std::string item) {
+                        return !item.compare(prop);
+                    }) == leftVarColNames.end()) {
+                return false;
+            }
```

Comment:
it's more clear to use for-loop or define a variable for std::find_if

Reply:
ACK.